File: haskell/pids/PitchRollAngle.hs
{--
  Pitch/roll angle PID-control algorithm for real and simulated flight
  controllers
 
  Copyright (C) 2024 Simon D. Levy
 
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, in version 3.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
--} 

{-# LANGUAGE DataKinds #-}
{-# LANGUAGE RebindableSyntax #-}

module PitchRollAngle where

import Language.Copilot
import Copilot.Compile.C99

import Demands
import State
import Utils

run kp dt target actual = demand where

    error = target - actual

    demand = kp * error

{--
  Demand is input as angles in degrees and output as angular velocities
  in degrees per second; roll-right / pitch-forward positive.
--}

pitchRollAngleController dt state demands = demands' where

  kp = 6

  rollDemand = run kp dt (roll demands) (phi state)

  pitchDemand = run kp dt (pitch demands) (theta state)

  demands' = Demands (thrust demands) rollDemand pitchDemand (yaw demands)

File: attic/webots/controllers/python/pids/pitch_roll_angle.py
from . pid import PID


class PitchRollAngleController:
    '''
       Demand is input as angles in degrees and output as angular velocities in
       degrees per second; roll-right / pitch-forward positive.
    '''

    def __init__(self):

        self.roll_pid = PID()

        self.pitch_pid = PID()

    def run(self, kp, phi, theta, dt, demands):

        t, roll, pitch, y = demands

        roll = self.run_axis(kp, self.roll_pid, roll, dt, phi)

        pitch = self.run_axis(kp, self.pitch_pid, pitch, dt, theta)

        return t, roll, pitch, y

    def run_axis(self, kp, pid, demand, dt, actual):

        return pid.run_p(kp, dt, demand, actual)
